src/Microsoft.VisualStudio.Composition/RuntimeExportProviderFactory+RuntimeExportProvider.cs

```diff
@@ -271,88 +267,124 @@ private object CreateExportFactory(RuntimePartLifecycleTracker importingPartTrac
+            private bool TryHandleGetExportProvider(RuntimeComposition.RuntimePart exportingRuntimePart, bool lazy, out object? exportProvider)
+            {
+                Requires.NotNull(exportingRuntimePart, nameof(exportingRuntimePart));
+
+                // Special case importing of ExportProvider
+                if (exportingRuntimePart.TypeRef.Equals(ExportProvider.ExportProviderPartDefinition.TypeRef))
+                {
+                    exportProvider = lazy ? () => this.NonDisposableWrapper.Value :
+                                                  this.NonDisposableWrapper.Value;
+
+                    return true;
+                }
+
+                exportProvider = null;
+                return false;
+            }
+
+            private PartLifecycleTracker GetOrCreateValue(RuntimeComposition.RuntimeImport import, RuntimeComposition.RuntimePart exportingRuntimePart, TypeRef originalPartTypeRef, TypeRef constructedPartTypeRef, RuntimePartLifecycleTracker? importingPartTracker)
             {
                 Requires.NotNull(import, nameof(import));
-                Requires.NotNull(export, nameof(export));
                 Requires.NotNull(exportingRuntimePart, nameof(exportingRuntimePart));
                 Requires.NotNull(originalPartTypeRef, nameof(originalPartTypeRef));
                 Requires.NotNull(constructedPartTypeRef, nameof(constructedPartTypeRef));
 
                 bool nonSharedInstanceRequired = !exportingRuntimePart.IsShared || import.IsNonSharedInstanceRequired;
                 Requires.Argument(importingPartTracker is object || !nonSharedInstanceRequired, nameof(importingPartTracker), "Value required for non-shared parts.");
                 RuntimePartLifecycleTracker? nonSharedPartOwner = nonSharedInstanceRequired && importingPartTracker!.IsNonShared && !import.IsExportFactory ? importingPartTracker : null;
-                PartLifecycleTracker partLifecycle = this.GetOrCreateValue(
+
+                return this.GetOrCreateValue(
                     originalPartTypeRef,
                     constructedPartTypeRef,
                     exportingRuntimePart.SharingBoundary,
                     import.Metadata,
                     nonSharedInstanceRequired,
                     nonSharedPartOwner);
-                var faultCallback = this.faultCallback;
+            }
+
+            /// <summary>
+            /// Called from <see cref="GetExportedValue(RuntimeComposition.RuntimeImport, RuntimeComposition.RuntimeExport, RuntimePartLifecycleTracker?, bool, out PartLifecycleTracker?)"/>
+            /// only, as an assisting method. See remarks.
+            /// </summary>
+            /// <remarks>
+            /// This method is separate from its one caller to avoid a csc.exe compiler bug
+            /// where it captures "this" in the closure for exportedValue, resulting in a memory leak
+            /// which caused one of our GC unit tests to fail.
```

Comment:
Is the compiler bug filed? Can we link to it from here?
Would a `static` local method suffice? Mostly just curious, but if it really only is meant to have the one caller, I tend to use local methods for things like that.

Reply:
This comment wasn't new, it was copied from prior code. I've inlined the method and added a comment referring to what I'm trying to avoid.